### PR TITLE
Add 2D loss map computation using pmap

### PR DESCRIPTION
**Pull request by haakon-e**
This PR adds the possibility to compute 2D maps of the loss function for a list of parameters using the config.jl file. 
The file `grid_search.jl` receives the config file and formulate all 2 members combinations in the parameter list and runs TC.jl for each parameter combinations and saves the outputs in specific folders.

The file `loss_map.jl` receives the path to the grid search output data and computes a loss map for each combinations of parameters from the TC.jl run done by `grid_search.jl`. The loss function is computed based on the config file for selected variables and with respect to LES and SCM. 

## Checklist before requesting a review / merging
- [x] I have formatted the code using `julia --project=.dev .dev/climaformat.jl .`
- [x] I have updated all test manifests using `julia --project .dev/up_deps.jl .` with Julia 1.7.0.
- [x] If core features are added, I have added appropriate test coverage.
- [x] If new functions and structs are added, they are documented through docstrings.

### CODE REVIEW

File: experiments/grid_search/output/entrainment_factor.detrainment_factor.20220318_kf5/config.jl [hunk 1]
```diff
@@ -0,0 +1,160 @@
+#= Custom calibration configuration file. =#
```
Comment (yairchn) [inline experiments/grid_search/output/entrainment_factor.detrainment_factor.20220318_kf5/config.jl:1]: we need to add to .gitignore the `output` folder

File: experiments/grid_search/runscript.sbatch [hunk 2]
```diff
@@ -0,0 +1,17 @@
+#!/bin/bash
```
Comment (yairchn) [inline experiments/grid_search/runscript.sbatch:1]: maybe we can change the file name to `grid_search.sbatch`

File: experiments/grid_search/loss_map.jl [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+    # compute loss, store in matrix: 
+    # (case, ens_i, p1, p2)
+    pval1 = parameters[first(param_names)]
+    pval2 = if length(param_names) == 1
+        nothing
+    else
+        parameters[last(param_names)]
+    end
+    n_pval1 = length(pval1)
+    n_pval2
```
Comment (yairchn) [inline experiments/grid_search/loss_map.jl:36]: `= length(pval2)` ?

File: experiments/grid_search/loss_map.jl [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+    loss = zeros((length(case_names), n_ens, n_pval1, n_pval2))
+
+    for p1_ind in 1:n_pval1
+        for p2_ind in 1:n_pval2
+            param_path = joinpath(sims_path, "$p")
```
Comment (yairchn) [inline experiments/grid_search/loss_map.jl:41]: is `p` defined above?

File: Project.toml [hunk 5]
```diff
@@ -46,7 +47,6 @@ ClimaCore = "0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8"
 CloudMicrophysics = "0.3"
 Dierckx = "0.5"
 Distributions = "0.25"
-EnsembleKalmanProcesses = "0.5"
```
Comment (ilopezgp) [inline Project.toml:49]: why do you take out this dependency?

File: experiments/grid_search/config.jl [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+function get_config()
+    config = Dict()
+    # Flags for saving output data
+    config["output"] = get_output_config()
+    # Define regularization of inverse problem
+    config["regularization"] = get_regularization_config()
+    # Define reference used in the inverse problem 
+    config["reference"] = get_reference_config(Bomex())
+    # Define reference models to use for validation
+    config["validation"] = get_reference_config(ValidateBomex())
```
Comment (ilopezgp) [inline experiments/grid_search/config.jl:30]: No need for validation in grid search.

File: experiments/grid_search/config.jl [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+    config["scm_parent_dir"] = ["scm_init"]
+    config["t_start"] = [4.0 * 3600]
+    config["t_end"] = [6.0 * 3600]
+    # Specify averaging intervals for covariance, if different from mean vector (`t_start` & `t_end`)
+    # config["Σ_t_start"] = [...]
+    # config["Σ_t_end"] = [...]
+    # If isnothing(config["batch_size"]), use all data per iteration
+    config["batch_size"] = nothing
+    return config
+end
```
Comment (ilopezgp) [inline experiments/grid_search/config.jl:157]: can be deleted.

File: experiments/grid_search/grid_search.jl [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+    using CalibrateEDMF.HelperFuncs
+    using CalibrateEDMF.DistributionUtils
+    using CalibrateEDMF.TurbulenceConvectionUtils
+    using Combinatorics
+    using TurbulenceConvection
+    tc = pkgdir(TurbulenceConvection)
+    include(joinpath(tc, "driver", "generate_namelist.jl"))
+
+    run_sims(t) = run_sims(t...)
+    function run_sims(param_values, case, j, nt)
```
Comment (ilopezgp) [inline experiments/grid_search/grid_search.jl:35]: typing and more descriptive two last arguments.

File: experiments/grid_search/grid_search.jl [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+            namelist_args = nt.namelist_args,
+            uuid = "$j",
+            les = get(namelist["meta"], "lesfile", nothing),
+        )
+    end
+end  # end @everywhere
+
+function grid_search(
+    sim_type = "reference",
+    root = pwd(),
```
Comment (ilopezgp) [inline experiments/grid_search/grid_search.jl:56]: config should be argument.

File: experiments/grid_search/loss_map.jl [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+@everywhere begin
+    using CalibrateEDMF
+    using CalibrateEDMF.HelperFuncs
+    using CalibrateEDMF.ReferenceStats
+    using CalibrateEDMF.LESUtils
+    using Combinatorics
+    import NCDatasets
+    const NC = NCDatasets
+    config_path = "config.jl"
+    include(config_path)
```
Comment (ilopezgp) [inline experiments/grid_search/loss_map.jl:14]: parse config as argument

File: experiments/grid_search/loss_map.jl [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+
+    function compute_loss_map(
+        # config_path,
+        sim_type = "reference",
+        sims_path = "output/20220328_bNQ/",
```
Comment (ilopezgp) [inline experiments/grid_search/loss_map.jl:19]: What is this default?

File: experiments/grid_search/loss_map.jl [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+    )
+        config = get_config()
```
Comment (ilopezgp) [inline experiments/grid_search/loss_map.jl:21]: Pass as argument.

File: experiments/grid_search/loss_map.jl [hunk 13]
```diff
@@ -0,0 +1,115 @@
...
+        n_ens = get_entry(config["grid_search"], "ensemble_size", nothing)
+        case_names = get_entry(config[sim_type], "case_name", nothing)
+
+        # param_paths = filter(isdir, readdir(sims_path, join=true))
+        parameters = get_entry(config["grid_search"], "parameters", nothing)
+        param_names = collect(keys(parameters))
+        NC.Dataset(joinpath(sims_path,"loss_hypercube.nc"),"c") do ds
```
Comment (ilopezgp) [inline experiments/grid_search/loss_map.jl:28]:
```suggestion
        NC.Dataset(joinpath(sims_path,"loss_slices_2D.nc"),"c") do ds
```
? It is not really a hypercube, just 2D slices of it.

File: experiments/grid_search/loss_map.jl [hunk 14]
```diff
@@ -0,0 +1,115 @@
...
+                nt = (; pval1, pval2, sims_path, group_name, config, sim_type)
+
+                loss_configs = vec(collect( Iterators.product(1:n_pval1, 1:n_pval2, 1:n_cases, 1:n_ens, [nt]) ))
+
+                @time begin
+                    sim_loss = pmap(compute_loss, loss_configs)
+                    loss = reshape(sim_loss, size(loss))
+                end
+                ensemble_member = LinRange(1, n_ens, n_ens)
+                group_root = ds.group[group_name]  # group is 'sorting_power.entrainment_factor'
```
Comment (ilopezgp) [inline experiments/grid_search/loss_map.jl:58]:
```suggestion
                group_root = ds.group[group_name]  # group is e.g. 'sorting_power.entrainment_factor'
```

File: experiments/grid_search/loss_map.jl [hunk 15]
```diff
@@ -0,0 +1,115 @@
...
+                ncvar = NC.defVar(group_root, param2,pval2,("param2",))
+                ncvar[:] = pval2
+                ncvar = NC.defVar(group_root, "loss_data", loss, ("param1", "param2", "case", "ensemble_member"))
+                ncvar[:,:,:,:] = loss
+            end
+        end  # do-block
+    end
+
+    compute_loss(t) = compute_loss(t...)
+    function compute_loss(p1_i, p2_j, case_k, ens_l, nt)
```
Comment (ilopezgp) [inline experiments/grid_search/loss_map.jl:81]: document function, more descriptive arguments

File: experiments/grid_search/loss_map.jl [hunk 16]
```diff
@@ -0,0 +1,115 @@
...
+        y_loss_names = if (y_ref_type == LES)
+            get_les_names(loss_names[case_k], filename)
+        else
+            loss_names[case_k]
+        end
+        y_full_case = get_profile(filename, y_loss_names, ti = t_start[case_k], tf = t_end[case_k], z_scm = z_scm)
+        g_full_case_i = get_profile(
+            nc_file, loss_names[case_k], ti = t_start[case_k], tf = t_end[case_k], z_scm = z_scm
+        )
+        sim_loss = compute_mse([g_full_case_i], y_full_case)[1]
```
Comment (ilopezgp) [inline experiments/grid_search/loss_map.jl:110]: Here you are concatenating variables without normalization, which is not really meaningful (e.g., concatenating temperature and ql means that you discard ql).

I think it is important to know: What landscape are you trying to understand? It seems to me like the appropriate landscape to look at is the one that the EKP actually sees.

**Comment by ilopezgp** [comment]
Please include an integration test as part of this PR, since none of the new code is currently tested.

File: experiments/grid_search/loss_map.jl (hunk 16 above)
Comment (haakon-e) [reply experiments/grid_search/loss_map.jl:110]: we leave this to another PR.

File: experiments/grid_search/config.jl (hunk 6 above)
Comment (haakon-e) [reply experiments/grid_search/config.jl:30]: we keep functionality for grid search over validation data if people desire to do so (it is trivial in terms of implementation). The `config.jl` file in this folder isn't intended to be used operationally. Instead you would be using a config file from some existing calibration procedure (modified to include `get_grid_search_config` function)

**Comment by ilopezgp** [comment]
Is this ready to be reviewed?

**Comment by yairchn** [comment]
> Is this ready to be reviewed? Buildkite seems to be failing.

yes, we will r

**Comment by ilopezgp** [comment]
Buildkite error: `ERROR: LoadError: IOError: open("grid_search_integration_test_config.jl", 0, 0): no such file or directory (ENOENT)`

File: Project.toml [hunk 17]
```diff
@@ -8,8 +8,10 @@ ArgParse = "c7e460c6-2fb9-53a9-8c5b-16f535851c63"
 ArtifactWrappers = "a14bc488-3040-4b00-9dc1-f6467924858a"
 AtmosphericProfilesLibrary = "86bc3604-9858-485a-bdbe-831ec50de11d"
 CLIMAParameters = "6eacf6c3-8458-43b9-ae03-caf5306d3d53"
+CairoMakie = "13f3f980-e62b-5c42-98c6-ff1f3baf88f0"
 ClimaCore = "d414da3d-4745-48bb-8d80-42e94e092884"
 CloudMicrophysics = "6a9e3e04-43cd-43ba-94b9-e8782df3c71b"
+Combinatorics = "861a8166-3701-5b0c-9a16-15d98fcdc6aa"
```
Comment (ilopezgp) [inline Project.toml:14]: Can you add a `[compat]` entry for CairoMakie and Combinatorics? We need compats to be able to publish versions.

File: experiments/grid_search/README.md [hunk 18]
```diff
@@ -0,0 +1,58 @@
+## Performing a grid search and calculating loss maps over pairs of parameters for the TurbulenceConvection.jl EDMF
+
+Authors: Haakon Ludvig Langeland Ervik, Yair Cohen
+
+Last updated : April 2022
+
+This tool relies on a `config.jl` used in a calbration process. To run this, the `config.jl` file may need to be modified with the following sub-dictionary:
```
Comment (ilopezgp) [inline experiments/grid_search/README.md:7]:
```suggestion
This tool relies on a `config.jl` used in a calbration process. To run this, the following sub-dictionary needs to be added to the `config.jl` file:
```

File: experiments/grid_search/README.md [hunk 19]
```diff
@@ -0,0 +1,58 @@
...
+```
+function get_grid_search_config()
+    config = Dict()
+    # Grid search is performed over each pair of parameters, across all specified values
```
Comment (ilopezgp) [inline experiments/grid_search/README.md:11]: From my point of view, showing a grid search example with 2 parameters can be misleading. User may interpret that if they add another parameter, they will get a cube (in fact, this is what grid search typically refers to). But my understanding is that if I were to add another parameter, you would only compute 2D slices of the cube, centered about the mean of the cube?

It would be enlightening to change the example in this README to 3 parameters, even if you end up running the example only with 2 for computational efficiency.

File: experiments/grid_search/README.md [hunk 20]
```diff
@@ -0,0 +1,58 @@
...
+```
+# grid search configuration
+config["grid_search"] = get_grid_search_config()
+```
+
+In the grid search config, 
+- `parameters` is a dictionary that specifies the parameters that will be used to generate 2D loss maps, along with the values for each parameter.
+- `ensemble_size` is an integer which specifies the number of times a specific simulation configuration is repeated (typically useful for stochastic models)
+- `output_root_dir` specifies the parent directory to a directory called `output`, in which you may store all your grid search output. A specific grid search run is stored in a folder identified by today's date + a random three-letter string `YYmmdd_abc`. The full output path is therefore `<output_root_dir>/output/YYmmdd_abc`.
+- `sim_type` is `"reference"` is grid search is run over the `reference` configurations or `"validation"` if it is run over the validation configurations.
```
Comment (ilopezgp) [inline experiments/grid_search/README.md:35]:
```suggestion
- `sim_type` selects which dataset to use for the map: choose `"reference"` to run a grid search over the `reference` configurations, or `"validation"` to run a grid search over the validation configurations.
```

File: experiments/grid_search/README.md [hunk 21]
```diff
@@ -0,0 +1,58 @@
...
+
+Running a grid search generates an output folder in which the raw `TurbulenceConvection.jl` data is stored in a systematic way, as well as the `config.jl` file used to generate it.
+If you are on the Caltech Central Cluster, you can run grid search by adding it to the schedule:
+
+  >> sbatch run_grid_search.sbatch path/to/config.jl
+
+In case some folders are missing due to an incomplete grid search, you can restart the grid search by
+
+  >> sbatch restart_grid_search.sbatch path/to/output/YYmmdd_abc
```
Comment (ilopezgp) [inline experiments/grid_search/README.md:44]: Awesome!

File: experiments/grid_search/grid_search.jl [hunk 22]
```diff
@@ -0,0 +1,236 @@
...
+    using CalibrateEDMF.DistributionUtils
+    using CalibrateEDMF.TurbulenceConvectionUtils
+    using Combinatorics
+    using TurbulenceConvection
+    tc = pkgdir(TurbulenceConvection)
+    include(joinpath(tc, "driver", "generate_namelist.jl"))
+
+    # parameter dictionary can consist of both scalar and vector parameters.
+    # flatten to only scalar parameters s.t. a vector parameter 'vec' becomes 'vec_{1}', 'vec_{2}', etc.
+    function flatten_vector_parameters(param_dict::Dict)
```
Comment (ilopezgp) [inline experiments/grid_search/grid_search.jl:21]: Can we leverage `https://github.com/CliMA/CalibrateEDMF.jl/blob/main/src/DistributionUtils.jl#L32` instead?

File: experiments/grid_search/grid_search.jl [hunk 23]
```diff
@@ -0,0 +1,236 @@
...
+                    out_dict["$(param)_{$j}"] = value[j]
+                end
+            else
+                out_dict[param] = value
+            end
+        end
+        return out_dict
+    end
+
+    function prepare_input_parameters(namelist::Dict, p1::S, p2::S, v1::Number, v2::Number) where {S <: String}
```
Comment (ilopezgp) [inline experiments/grid_search/grid_search.jl:35]:
```suggestion
    function get_parameter_pairs(namelist::Dict, p1::S, p2::S, v1::Number, v2::Number) where {S <: String}
```
?

File: experiments/grid_search/grid_search.jl (hunk 22 above)
Comment (yairchn) [reply experiments/grid_search/grid_search.jl:21]: @haakon-e  this is your addition for stochastic closures, what do you think?

**Comment by ilopezgp** [comment]
The CI is still breaking. You will also have to rebase after #331 is merged. Sorry about that, but we need to be able to train on deep convection cases ASAP.

**Comment by ilopezgp** [comment]
bors try

File: experiments/grid_search/grid_search.jl (hunk 22 above)
Comment (haakon-e) [reply experiments/grid_search/grid_search.jl]: I removed the function entirely. The method you referenced @ilopezgp is too specific for what I needed, so I found some neat (almost) one-liners to do the same stuff.

File: src/HelperFuncs.jl [hunk 24]
```diff
@@ -224,13 +224,15 @@ Given directory to standard LES or SCM output, fetch path to stats file.
...
+        stat_files = glob(relpath(abspath(nc_path)))
+        @assert length(stat_files) == 1 "$(length(stat_files)) stats files found with paths $nc_path"
         return stat_files[1]
     end
     try
-        stat_files = glob(relpath(abspath(joinpath(stats, "*.nc"))))
-        @assert length(stat_files) == 1 "$(length(stat_files)) stats files found with paths $stat_files"
+        nc_path = joinpath(stats, "*.nc")
+        stat_files = glob(relpath(abspath(nc_path)))
+        @assert length(stat_files) == 1 "$(length(stat_files)) stats files found with paths $nc_path"
```
Comment (ilopezgp) [inline src/HelperFuncs.jl:235]: Does this fix #311 @haakon-e ?

File: src/ReferenceStats.jl [hunk 25]
```diff
@@ -136,15 +147,28 @@ Base.@kwdef struct ReferenceStatistics{FT <: Real}
...
 
 "Returns full dimensionality of the ReferenceStatistics, before latent space encoding"
 full_length(RS::ReferenceStatistics) = length(RS.y_full)
+pca_length(RS::ReferenceStatistics, case_ind) = RS.ndof_case[case_ind]
+full_length(RS::ReferenceStatistics, case_ind) = RS.ndof_full_case[case_ind]
+
+pca_inds(RS::ReferenceStatistics, case_ind) = case_inds(RS.ndof_case, case_ind)
+full_inds(RS::ReferenceStatistics, case_ind) = case_inds(RS.ndof_full_case, case_ind)
+
+case_inds(ndofs::Vector{IT}, case_ind::IT) where {IT <: Integer} = begin
```
Comment (ilopezgp) [inline src/ReferenceStats.jl:166]: Can you add a docstring for this function? Also, are you calling `ndofs` the number of cases? That should refer to degrees of freedom per case.

Comment (ilopezgp) [reply src/ReferenceStats.jl:166]: The fact that case_ind is an input to a function called case_inds is also very confusing.

**Comment by ilopezgp** [comment]
The output of the grid search on buildkite at the moment just includes a few blue dots, are we supposed to see a loss map?

**Comment by haakon-e** [comment]
> The output of the grid search on buildkite at the moment just includes a few blue dots, are we supposed to see a loss map?

I would have to expand the test to more parameters to get a loss map.
You get scatter plots when at least one of the parameters only has one value that we run grid search over.

**Comment by ilopezgp** [comment]
> > The output of the grid search on buildkite at the moment just includes a few blue dots, are we supposed to see a loss map?
> 
> I would have to expand the test to more parameters to get a loss map. You get scatter plots when at least one of the parameters only has one value that we run grid search over.

You can balance the load if you avoid computation elsewhere. Right now you are computing the map based on three different TC runs, you can just use two for example. You can also avoid the ensemble business (`config["ensemble_size"] = 2`). But the test should have at least 2 parameter values per parameter and 3 parameters to test the basic functionality.

File: src/HelperFuncs.jl (hunk 24 above)
Comment (haakon-e) [reply src/HelperFuncs.jl:235]: yes.

File: src/ReferenceStats.jl (hunk 25 above)
Comment (haakon-e) [reply src/ReferenceStats.jl:166]: `ndofs` is the degrees of freedom per case (the vector has length equal to the number of cases). I renamed the function to reflect that in principle, it is more general than my specific use-case, and added docs.

**Comment by haakon-e** [comment]
bors try
